n_beats/model_keras.py

```diff
@@ -0,0 +1,296 @@
+                            color=['b'] * self.backcast_length + ['g'] * self.forecast_length)
+                plt.scatter(list(range(len(x_y) - self.forecast_length, len(x_y))),
+                            validation_predictions[0, :].flatten(), color=['r'] * self.forecast_length)
+                plt.legend(['backcast', 'forecast', 'predictions of forecast'])
+                if not os.path.exists('results/test'):
+                    os.makedirs('results/test')
+                plt.savefig('results/test/' + str(step) + '.png')
+                plt.close()
+
+        self.nbeats.save('results/n_beats_model_20_small.h5')
+
+        predictions = self.nbeats.predict(x)
+        x_y = np.concatenate([x[100, :], y[100, :]], axis=-1).flatten()
+        plt.plot(list(range(self.backcast_length)), x[100, :].flatten(), color='b')
+        plt.plot(list(range(len(x_y) - self.forecast_length, len(x_y))), y[100, :].flatten(), color='g')
+        plt.plot(list(range(len(x_y) - self.forecast_length, len(x_y))), predictions[100, :].flatten(), color='r')
+        plt.scatter(range(len(x_y)), x_y.flatten(), color=['b'] * self.backcast_length + ['g'] * self.forecast_length)
+        plt.scatter(list(range(len(x_y) - self.forecast_length, len(x_y))), predictions[100, :].flatten(),
+                    color=['r'] * self.forecast_length)
+        plt.legend(['backcast', 'forecast', 'predictions of forecast'])
+        plt.savefig('results/' + str(nb_epochs) + '.png')
+
+        validation_predictions = self.nbeats.predict(x_test)
+        x_y = np.concatenate([x_test[10, :], y_test[10, :]], axis=-1).flatten()
+        plt.figure()
+        plt.plot(list(range(self.backcast_length)), x_test[10, :].flatten(), color='b')
+        plt.plot(list(range(len(x_y) - self.forecast_length, len(x_y))), y_test[10, :].flatten(), color='g')
+        plt.plot(list(range(len(x_y) - self.forecast_length, len(x_y))), validation_predictions[10, :].flatten(),
+                 color='r')
+        plt.scatter(range(len(x_y)), x_y.flatten(), color=['b'] * self.backcast_length + ['g'] * self.forecast_length)
+        plt.scatter(list(range(len(x_y) - self.forecast_length, len(x_y))), validation_predictions[10, :].flatten(),
+                    color=['r'] * self.forecast_length)
+        plt.legend(['backcast', 'forecast', 'predictions of forecast'])
+        plt.savefig('results/test/' + str(nb_epochs) + '.png')
+
+        print(get_accuracy(y_test, validation_predictions))
+
+
+def linear_space(length, fwd_looking=True):
+    if fwd_looking:
+        t = np.linspace(0, length - 1, length)
+    else:
+        t = np.linspace(-length, -1, length)
+    t = t / length
+    return t
+
+
+def seasonality_model(thetas, length, is_forecast):
+    p = thetas.shape[-1]
+    p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
+    t = linear_space(length, fwd_looking=is_forecast)
+    s1 = np.stack([np.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)  # H/2-1
```

Comment:
you can use tf.cos instead of np.cos for all.

Something like this would work (to be checked!)

```
p = thetas.get_shape().as_list()[-1]
p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
s1 = tf.stack([tf.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)  # H/2-1
s2 = tf.stack([tf.sin(2 * np.pi * i * t) for i in range(p2)], axis=0)
 S = tf.concat([s1, s2], axis=0)
return tf.matmul(thetas, S)
```

Reply:
Corrected